pkg/server/plugin/upstreamauthority/gcpcas/gcpcas.go

```diff
@@ -0,0 +1,393 @@
+package gcpcas
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	pcaapi "cloud.google.com/go/security/privateca/apiv1beta1"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/pkg/server/plugin/upstreamauthority"
+	"github.com/spiffe/spire/proto/spire/common/plugin"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	privatecapb "google.golang.org/genproto/googleapis/cloud/security/privateca/v1beta1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+const (
+	// The name of the plugin
+	pluginName    = "gcp_cas"
+	publicKeyType = "PUBLIC KEY"
+)
+
+func makeError(code codes.Code, format string, args ...interface{}) error {
+	return status.Errorf(code, "GCPCAS: "+format, args...)
+}
+
+// BuiltIn constructs a catalog Plugin using a new instance of this plugin.
+func BuiltIn() catalog.Plugin {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.Plugin {
+	return catalog.MakePlugin(pluginName, upstreamauthority.PluginServer(p))
+}
+
+type CertificateAuthoritySpec struct {
+	Project    string `hcl:"project_name"`
+	Location   string `hcl:"region_name"`
+	LabelKey   string `hcl:"label_key"`
+	LabelValue string `hcl:"label_value"`
+}
+
+func (spec *CertificateAuthoritySpec) caParentPath() string {
+	return path.Join("projects", spec.Project, "locations", spec.Location)
+}
+
+type Config struct {
+	RootSpec         CertificateAuthoritySpec   `hcl:"root_cert_spec,block"`
+	TrustBundleSpecs []CertificateAuthoritySpec `hcl:"trust_bundle_cert_spec,block"`
+}
+
+type CAClient interface {
+	CreateCertificate(ctx context.Context, req *privatecapb.CreateCertificateRequest) (*privatecapb.Certificate, error)
+	LoadCertificateAuthorities(ctx context.Context, spec CertificateAuthoritySpec) ([]*privatecapb.CertificateAuthority, error)
+}
+
+type gcpCAClient struct {
+	pcaClient *pcaapi.CertificateAuthorityClient
+}
+
+func (client *gcpCAClient) CreateCertificate(ctx context.Context, req *privatecapb.CreateCertificateRequest) (*privatecapb.Certificate, error) {
+	return client.pcaClient.CreateCertificate(ctx, req)
+}
+func (client *gcpCAClient) LoadCertificateAuthorities(ctx context.Context, spec CertificateAuthoritySpec) ([]*privatecapb.CertificateAuthority, error) {
+	// https://pkg.go.dev/cloud.google.com/go/security/privateca/apiv1beta1#CertificateAuthorityClient.ListCertificateAuthorities
+	allCerts := make([]*privatecapb.CertificateAuthority, 0)
```

Comment:
nit: Unless you are specifying a length or capacity, or need to disambiguate between a nil and empty slice for some reason, we prefer the following syntax for declaration:
```suggestion
	var allCerts []*privatecapb.CertificateAuthority
```
`make` usage requires extra scrutiny by code readers to ensure the parameters are correct so it's better to avoid it if there isn't a compelling reason like the ones listed above.